internal/appsec/waf/waf_test.go

```diff
@@ -406,6 +407,105 @@ func TestRunError(t *testing.T) {
 	}
 }
 
+func TestMetrics(t *testing.T) {
+	rules := `
+{
+  "version": "2.1",
+  "metadata": {
+    "rules_version": "1.2.7"
+  },
+  "rules": [
+	{
+	  "id": "valid-rule",
+	  "name": "Unicode Full/Half Width Abuse Attack Attempt",
+	  "tags": {
+		"type": "http_protocol_violation"
+	  },
+	  "conditions": [
+		{
+		  "parameters": {
+			"inputs": [
+			  {
+				"address": "server.request.uri.raw"
+			  }
+			],
+			"regex": "\\%u[fF]{2}[0-9a-fA-F]{2}"
+		  },
+		  "operator": "match_regex"
+		}
+	  ],
+	  "transformers": []
+	},
+	{
+	  "id": "missing-tags-1",
+	  "name": "Unicode Full/Half Width Abuse Attack Attempt",
+	  "conditions": [
+	  ],
+	  "transformers": []
+	},
+	{
+	  "id": "missing-tags-2",
+	  "name": "Unicode Full/Half Width Abuse Attack Attempt",
+	  "conditions": [
+	  ],
+	  "transformers": []
+	},
+	{
+	  "id": "missing-name",
+	  "tags": {
+		"type": "http_protocol_violation"
+	  },
+	  "conditions": [
+	  ],
+	  "transformers": []
+	}
+  ]
+}
+`
+	waf, err := NewHandle([]byte(rules))
+	require.NoError(t, err)
+	defer waf.Close()
+	t.Run("RulesetInfo", func(t *testing.T) {
+		rInfo := waf.RulesetInfo()
+		require.Equal(t, uint16(3), rInfo.Failed)
+		require.Equal(t, uint16(1), rInfo.Loaded)
+		require.Equal(t, 2, len(rInfo.Errors))
+		require.Equal(t, "1.2.7", rInfo.Version)
+		for key, arr := range rInfo.Errors {
+			arr := arr.([]interface{})
+			switch key {
+			case "missing key 'tags'":
+				require.Equal(t, 2, len(arr))
+				require.Equal(t, "missing-tags-1", arr[0])
+				require.Equal(t, "missing-tags-2", arr[1])
+				break
+			case "missing key 'name'":
+				require.Equal(t, 1, len(arr))
+				require.Equal(t, "missing-name", arr[0])
+				break
+			default:
+				t.Error("unexpected key in RulesetInfo.Errors")
+			}
+		}
+	})
+
+	t.Run("RunDuration", func(t *testing.T) {
+		wafCtx := NewContext(waf)
+		require.NotNil(t, wafCtx)
+		defer wafCtx.Close()
+		// Craft matching data to force work on the WAF
+		data := map[string]interface{}{
+			"server.request.uri.raw": "\\%uff00",
+		}
+		matches, err := wafCtx.Run(data, time.Second)
+		require.NoError(t, err)
+		require.NotNil(t, matches)
+		// Make sure that WAF runtime was set
+		require.Greater(t, wafCtx.TotalRuntime(), uint64(0), "wafCtx runtime metric is not set")
+		require.LessOrEqual(t, wafCtx.TotalRuntime(), uint64(time.Second), "wafCtx runtime metric is incorrect")
```

Comment:
We can do better and more robust by timing the call to `wafCtx.Run(data, time.Second)` and then asserting that the `wafCtx.TotalRuntime()` is <= to what you timed from here.